### PR TITLE
TCOTA-542 Implemented drag and drop for treatment questions

### PR DESCRIPTION
**Pull request by rmathis**

### CODE REVIEW

**Review by machristie — CHANGES_REQUESTED** [review]
The look here is really good @rmathis . As is, though, this only updates the order of questions within a page of questions. See my comments. I think operating on the entire questions array will work better.

My advice on how to handle the question reordering is to:

- change the "assessment/questions" getter to order the questions by questionOrder
- in handleQuestionOrderChange, create a copy of `this.questions`
- reorder the copy based on oldIndex and newIndex (much like you are already doing)
- loop over the copy and call the updateQuestions action for each one but set the questionOrder of each question to its index in the copied list from the previous step

Actually, you're already doing those last two in handleQuestionOrderChange. Just need to operate on the entire array.

Also, just a heads up, when a page break is created, it gets added at an index in the list of questions. See action "assessment/createQuestionAtIndex". After the page break question is inserted in the list of questions, the questionOrder should be set for all questions based on its index in the list.

File: src/main/javascript/src/views/assignment/TerracottaBuilder.vue [hunk 1]
```diff
@@ -36,26 +36,33 @@
...
               <div :key="questionPage.key">
                 <v-expansion-panels
-                  class="v-expansion-panels--outlined"
                   flat
                   accordion
+                  outlined
+                  class="v-expansion-panels--outlined"
                   :key="questionPage.key"
                 >
+                <draggable v-model="questionPage.questions" handle=".dragger"
```
[the remark this answers is not in the document]
Comment (rmathis) [reply src/main/javascript/src/views/assignment/TerracottaBuilder.vue]: Fixed

File: src/main/javascript/src/views/assignment/TerracottaBuilder.vue [hunk 2]
```diff
@@ -234,6 +244,17 @@ export default {
         console.error(error);
       }
     },
+    async handleQuestionOrderChange(event, page) {
+      const { questions } = page;
+      let list = [...questions.map(q => ({...q}))];
+      const movedItem = list.splice(event.oldIndex, 1)[0];
+      list.splice(event.newIndex, 0, movedItem);
+      list.forEach((q, idx) => {
+        q.questionOrder = idx;
```
[the remark this answers is not in the document]
Comment (rmathis) [reply src/main/javascript/src/views/assignment/TerracottaBuilder.vue]: Fixed

**Comment by rmathis** [comment]
Hi Marcus, I needed to change how I was calculating oldIndex, but otherwise your advice was sound. Thanks!

File: src/main/javascript/src/views/assignment/TerracottaBuilder.vue [hunk 3]
```diff
@@ -222,6 +232,7 @@ export default {
       updateQuestion: "assessment/updateQuestion",
       deleteQuestion: "assessment/deleteQuestion",
       updateAnswer: "assessment/updateAnswer",
+      orderQuestions: "assessment/orderQuestions"
```
Comment (machristie) [inline src/main/javascript/src/views/assignment/TerracottaBuilder.vue:235]: I'm not seeing orderQuestions implementation, or where it is used?

**Review by machristie — CHANGES_REQUESTED** [review]
@rmathis this looks much better! I like it. I just have a question about orderQuestions, see comment below.

File: src/main/javascript/src/views/assignment/TerracottaBuilder.vue (hunk 3 above)
Comment (rmathis) [reply src/main/javascript/src/views/assignment/TerracottaBuilder.vue]: That was a mistake, I didn't mean to leave it in. I was debating lifting the logic up into a vuex action but ended up going another route.

Fixed

**Review by machristie — APPROVED** [review]
Nice, thanks @rmathis

**Pull request merged**